core/drivers/crypto/se050/adaptors/apis/apdu.c

```diff
@@ -0,0 +1,685 @@
+		goto label_public;
+
+	public_keylen = key_pair->x_len + key_pair->y_len + 1;
+	if (public_keylen > sizeof(buffer)) {
+		EMSG("small buffer");
+		goto exit;
+	}
+	memcpy(public_key, key_pair->x, key_pair->x_len);
+	memcpy(public_key + key_pair->x_len, key_pair->y, key_pair->y_len);
+
+	status = Se05x_API_WriteECKey(s_ctx, &policy,
+				      SE05x_MaxAttemps_UNLIMITED,
+				      k_object->keyId,
+				      (SE05x_ECCurve_t)key_pair->curve,
+				      key_pair->d, key_pair->d_len,
+				      buffer, public_keylen, type,
+				      kSE05x_KeyPart_Pair);
+
+	ret = status == SM_OK ? kStatus_SSS_Success : kStatus_SSS_Fail;
+	goto exit;
+
+label_public:
+	if (k_object->objectType != kSSS_KeyPart_Public)
+		goto label_private;
+
+	public_keylen = key_pub->x_len + key_pub->y_len + 1;
+	if (public_keylen > sizeof(buffer)) {
+		EMSG("small buffer");
+		goto exit;
+	}
+	memcpy(public_key, key_pub->x, key_pub->x_len);
+	memcpy(public_key + key_pub->x_len, key_pub->y, key_pub->y_len);
+
+	status = Se05x_API_WriteECKey(s_ctx, &policy,
+				      SE05x_MaxAttemps_UNLIMITED,
+				      k_object->keyId,
+				      (SE05x_ECCurve_t)key_pub->curve,
+				      NULL, 0, buffer, public_keylen,
+				      type, kSE05x_KeyPart_Public);
+	ret = status == SM_OK ? kStatus_SSS_Success : kStatus_SSS_Fail;
+	goto exit;
+
+label_private:
+	if (k_object->objectType != kSSS_KeyPart_Private)
+		goto exit;
+
+	status = Se05x_API_WriteECKey(s_ctx, &policy,
+				      SE05x_MaxAttemps_UNLIMITED,
+				      k_object->keyId,
+				      (SE05x_ECCurve_t)key_pair->curve,
+				      key_pair->d, key_pair->d_len,
+				      NULL, 0, type, kSE05x_KeyPart_Private);
+	ret = status == SM_OK ? kStatus_SSS_Success : kStatus_SSS_Fail;
+	goto exit;
+
+exit:
+	return ret;
+}
+
+sss_status_t se050_key_store_get_ecc_key_bin(sss_se05x_key_store_t *store,
+					     sss_se05x_object_t *k_object,
+					     uint8_t *key, size_t *key_len)
+{
+	Se05xSession_t *s_ctx = NULL;
+	sss_cipher_type_t cipher_type = kSSS_CipherType_NONE;
+	sss_status_t ret = kStatus_SSS_Fail;
+	smStatus_t status = SM_NOT_OK;
+	uint8_t *key_buf = NULL;
+	size_t key_buflen = 0;
+
+	if (!store || !store->session || !k_object || !key || !key_len)
+		return kStatus_SSS_Fail;
+
+	s_ctx = &store->session->s_ctx;
+	cipher_type = k_object->cipherType;
+
+	switch (cipher_type) {
+	case kSSS_CipherType_EC_NIST_P:
+	case kSSS_CipherType_EC_NIST_K:
+	case kSSS_CipherType_EC_BRAINPOOL:
+	case kSSS_CipherType_EC_BARRETO_NAEHRIG:
+	case kSSS_CipherType_EC_MONTGOMERY:
+	case kSSS_CipherType_EC_TWISTED_ED:
+		add_ecc_header(key, &key_buf, &key_buflen, k_object->curve_id);
+		status = Se05x_API_ReadObject(s_ctx, k_object->keyId,
+					      0, 0, key_buf, key_len);
+		if (status != SM_OK)
+			goto exit;
+
+		*key_len += key_buflen;
+
+		/* return only the binary data */
+		key_buflen = *key_len;
+		get_ecc_raw_data(key, &key_buf, &key_buflen,
+				 k_object->curve_id);
+		*key_len = key_buflen;
+		memcpy(key, key_buf, key_buflen);
```

Comment:
Is there a possibility of key_buflen exceeding the length of key buffer here ? Check to ensure this would be good.

Reply:
ok. need to review the ECC side to make sure we pass the max key_len

Reply:
ok